### ISSUE TITLE
Dependencies

### ISSUE THREAD

**Issue by rafaqz** [JuliaImages/ImageMorphology.jl#40]
A lot of the methods in this package are widely useful outside of image manipulation. 

In ecology and geospatial fields there are uses for methods like `label_components` and `feature_distance` but dependencies like ColorVectorSpace.jl and ImageCore.jl are a lot to include.

Looking through the source, I cant actually find much that requires these dependencies. Could `GenericGrayImage` just be `AbstractArray` ? What other methods require ColorVectorSpace.jl and ImageCore.jl?

See:
https://github.com/EcoJulia/NeutralLandscapes.jl/issues/14

**Comment by mkborregaard** [comment]
There has been a similar discussion over at ImageFiltering JuliaImages/ImageFiltering.jl#42 and JuliaImages/ImageFiltering.jl#51 and it seems that @timholy agrees this would make sense too. There are so many excellent general array functions in the image ecosystem that could benefit the entire ecosystem immensely if they came in a clearer more light-weight format

**Comment by rafaqz** [comment]
As with Imagefiltering.jl, moving a lot of this code somewhere without `Image-` in the name would make it more discoverable, and something more people are likely to contribute to.

**Comment by rafaqz** [comment]
`coords_spatial` is the only method form `ImageCore.jl`, used only in `dilation_and_erosion.jl`. Removing ColorVectorSpace.jl would mean users would have to import it manually for colors to work.

Swapping `GenericGrayImage` to `AbstractArray` and removing the types in `convexhull` (which included `Gray{Bool}`) are the only other changes required.

**Comment by rafaqz** [comment]
I would put in a PR, except the traits dependency in ImageCore.jl is blocking this being possible - they would need to move somewhere - ArrayInterface.jl?

**Comment by juliohm** [comment]
I agree that we should try to generalize and split things out in terms of AbstractArray. @johnnychen94 can provide more useful feedback on the future plans. It is been a while since the last time I contributed to the Images.jl stack and I know a lot has changed since then.

**Comment by johnnychen94** [comment]
This is a good proposal.

Except for the maintainability, a similar relationship between Distances.jl and ImageDistances.jl is an ideal solution; people outside of the JuliaImages community could directly use and extend a "Morphology.jl" package, and JuliaImages keeps maintaining the colorant part of it.

Maintainability is the only issue here; we don't have many maintainers here to split everything apart to get the least dependencies. (Mostly it's me and @timholy doing it).

If there's someone giving a clear fork of ImageMorphology without the ImageCore.jl stack, I'll be very happy to see ImageMorphology built on top it.

**Comment by johnnychen94** [comment]
This is a good proposal.

Except for the maintainability, a similar relationship between Distances.jl and ImageDistances.jl is an ideal solution; people outside of the JuliaImages community could directly use and extend a "Morphology.jl" package, and JuliaImages keeps maintaining the colorant part of it.

Maintainability is the only issue here; we don't have many maintainers here to split everything apart to get the least dependencies. (Mostly it's me and @timholy doing it).

If there's someone giving a clear fork of ImageMorphology without the ImageCore.jl stack, I'll be very happy to see ImageMorphology built on top it.

**Comment by rafaqz** [comment]
Great. I understand the maintenance burden involved here, and don't have a huge amount available myself. But hopefully this would eventually spread the burden a little wider. 

We could register a very minimally changed version of this package as Morphology.jl with type restrictions eased and without the `coords_spatial` method. Instead we can define `coords`. `ImageMorphology can override `coords` for `AbstractArray{Color}` (if I understand how Images workds correctly) and forward to `coords_spatial`, and otherwise be a nearly empty package.

The next question is what would be a good location/org for Morphology.jl? It could be the same place as MapFilter.jl/Stencils.jl etc suggested in related issues (I would also be good to have those package separated out in a similar way).

**Comment by timholy** [comment]
I'm supportive of this. Hold for just a little while? ImageCore may soon be based on ArrayInterface (https://github.com/JuliaImages/ImageCore.jl/issues/112#issuecomment-765332944) but there will likely be a few things to sort out. I'd like to see ArrayInterface move to JuliaArrays first, and I haven't gone beyond https://github.com/SciML/ArrayInterface.jl/issues/121 in terms of checking whether ArrayInterface is in adequate shape to be included in ImageCore.

**Comment by rafaqz** [comment]
> I'd like to see ArrayInterface move to JuliaArrays first,

That would be great. I thought it was a stretch to propose moving array traits from ImageCore to SciML. But the images traits could live there if it was in JuliaArrays, and `coords_spatial` could stay.

I think we have a short term solution so we can wait, but I think the aim is to not to duplicates of these algorithms around.

**Comment by rafaqz** [comment]
@timholy an @johnnychen94 I'll put together Morphology.jl and write the PR to depend on it here once the ImageCore.jl traits are sorted. Reading them I'm not sure I can do that myself,  but I'm guessing that's possible now that ArrayInterface.jl has moved. I'm happy to maintain Morphology.jl as well so this isn't adding a burden here.

**Comment by rafaqz** [comment]
Just checking in on this, and where the ImageCore.jl traits are at now?

**Comment by johnnychen94** [comment]
Here's the plan with respect to the lazy loading(https://github.com/JuliaImages/Images.jl/issues/1009) and monorepo (https://github.com/JuliaImages/Images.jl/issues/898) stuff:

- the types and interface will go to "MorphologyInterface.jl" -- mainly the structuring element stuff so far
- the main algorithm codebase will be a standalone package "Morphology.jl" without ImageCore dependency. I still plan to depend on FixedPointNumbers because that's a core number type we used across the entire JuliaImages ecosystem and because that's super lightweight -- 0.004 seconds in my machine.
- the current ImageMorphology.jl will be migrated to Images.jl

**Comment by johnnychen94** [comment]
![image](https://user-images.githubusercontent.com/8684355/169639334-ee927197-7277-4c8d-8693-555b1c2700ff.png)

**Comment by johnnychen94** [comment]
Hmmm, I changed my mind wrt monorepo arch. Let's just get everything back to Images.jl
